106-auto-rebalance-on-imbalanced-clusters.md

```diff
@@ -0,0 +1,524 @@
+    RebalanceOnAnomalyDetection->>RebalanceOnScaleDown: execute scale down
+    RebalanceOnScaleDown-->>RebalanceOnScaleUp: then scale up
+    RebalanceOnScaleUp-->>RebalanceOnAnomalyDetection: then full rebalance
+  end
+  RebalanceOnAnomalyDetection->>Idle: scale up complete with no queued items
+```
+* from **RebalanceOnAnomalyDetection**:
+  * **RebalanceOnAnomalyDetection**: if another goal violation was detected while the first one ended. If a scale down and scale up is also queued up then they will execute first.
+  * **RebalanceOnScaleUp**: if a rebalancing on scale up is queued and will run if there is no other rebalancing scale down in queue. If a rebalancing scale down is in queue then it will be executed first.
+  * **RebalanceOnScaleDown**: if a scale down operation was requested, it will run once the full rebalance is completed
+  * **Idle**: if full rebalance was requested, it was executed and completed successfully or failed.
+
+On each reconciliation, the following process will be used:
+
+```mermaid
+flowchart TB
+  A[KafkaClusterCreator] --creates--> B[KafkaCluster]
+  B -- calls --> D[KafkaAutoRebalancingReconciler.reconcile]
+  D -- check for ConfigMap with goal-violation prefix --> E{if config map present?}
+  D -- if rebalance in progress --> F[moves a;; unfixed anomalies in list to fixed anomalies list]
+  E -- yes --> G[Trigger auto-rebalance]
+  E -- no --> H[No operation]
+```
+
+1. The `KafkaClusterCreator` creates the `KafkaCluster` instance.
+2. The `KafkaAutoRebalancingReconciler.reconcile()` will then check if there was any ConfigMap created with name  `goal-violation-map` and whether the `anomaly_list` list is empty or not, then the `full` rebalance(imbalance mode) would be performed if the list is not empty.
+If the list is not empty, then we trigger the `imbalance` mode (full rebalance). 
+3. In case, a rebalance is already ongoing and more anomalies are detected, then the operator will just ignore the new anomalies and delete all the anomalies from the `anomaly_list` in the ConfigMap.
+
+The `KafkaAutoRebalancingReconciler.reconcile()` loads the `Kafka.status.autoRebalance` content:
+
+* `state`: is the FSM state.
+* `lastTransitionTime`: when the transition to that state happened.
+* `modes`: sets the mode as `imbalance`
+
+The FSM is initialized based on the `state` field.
+
+Let's see what happens during the auto-rebalancing process when the FSM starts from the **Idle** state and transitions to **RebalanceOnAnomalyDetection**
+
+#### Idle
+
+This state is set since the beginning when a `Kafka` custom resource is created with the `spec.cruiseControl.autoRebalance` field.
+It is also the end state of a previous successfully completed or failed auto-rebalancing.
+In case of successful completion, once the rebalance moves to `Ready` state, we will delete the KafkaRebalance and move the empty the `anomaly_list` then update the `auto-rebalance` state to `Idle`.
```

Comment:
```suggestion
In case of successful completion, once the rebalance moves to `Ready` state, we will delete the `KafkaRebalance` and move the empty the `anomaly_list` then update the `auto-rebalance` state to `Idle`.
```